### PR TITLE
Accept `Path` arguments in `Compress::Zip`

### PR DESCRIPTION
**Pull request by didactic-drunk**
Split from #10866 as requested

### CODE REVIEW

**Review by asterite — APPROVED** [review]

File: src/compress/zip/checksum_reader.cr [hunk 1]
```diff
@@ -5,14 +5,14 @@ module Compress::Zip
   private class ChecksumReader < IO
     getter crc32 = ::Digest::CRC32.initial
 
-    def initialize(@io : IO, @filename : String, verify @expected_crc32 : UInt32? = nil)
+    def initialize(@io : IO, @filename : Path | String, verify @expected_crc32 : UInt32? = nil)
```
Comment (straight-shoota) [inline src/compress/zip/checksum_reader.cr:8]: I don't think there's much value in preserving the type union into the instance variable. Let's convene on a single type here (probably `String`?).

File: src/compress/zip/file.cr [hunk 2]
```diff
@@ -57,21 +57,21 @@ class Compress::Zip::File
...
     zip = new filename
     yield zip ensure zip.close
   end
 
   # Returns the entry that has the given filename, or
   # raises `KeyError` if no such entry exists.
-  def [](filename : String) : Entry
-    self[filename]? || raise(KeyError.new("Missing zip entry: #{filename}"))
+  def [](filename : Path | String) : Entry
+    self[filename.to_s]? || raise(KeyError.new("Missing zip entry: #{filename}"))
```
Comment (straight-shoota) [inline src/compress/zip/file.cr:68]: `to_s` shouldn't be necessary here since `#[]?` accepts the same types?

**Review by straight-shoota — APPROVED** [review]

File: src/compress/zip/checksum_reader.cr [hunk 3]
```diff
@@ -12,7 +12,7 @@ module Compress::Zip
       read_bytes = @io.read(slice)
       if read_bytes == 0
         if (expected_crc32 = @expected_crc32) && crc32 != expected_crc32
-          raise Compress::Zip::Error.new("Checksum failed for entry #{@filename} (expected #{expected_crc32}, got #{crc32}")
+          raise Compress::Zip::Error.new("Checksum failed for entry #{@filename.to_s} (expected #{expected_crc32}, got #{crc32}")
```
Comment (Sija) [inline src/compress/zip/checksum_reader.cr:15]: That's redundant.

```suggestion
          raise Compress::Zip::Error.new("Checksum failed for entry #{@filename} (expected #{expected_crc32}, got #{crc32}")
```

**Review by beta-ziliani — APPROVED** [review]
Thanks!

**Pull request merged**